SmartPointers/shared_pointer.hpp

```diff
@@ -1,80 +1,173 @@
-
-#ifndef shared_pointer_hpp
-#define shared_pointer_hpp
-
-using namespace std;
-#include <iostream>
-template<class T>
-class SharedPointer
-{
-private:
-    T* m_ptr;
-    unsigned int* m_count;
-public:
-     SharedPointer():m_ptr(nullptr), m_count(new unsigned int(0))
+
+
+
+
+template<typename T>
+class SharedPointer{
+
+    public:
+
+    //Constructor
+    SharedPointer():m_ptr(nullptr), count(new int(1))
     {
-        
+
     }
 
-    ~SharedPointer(){
-        //cleanup();
-        
+    //Paramatized Constructor
+     SharedPointer(T *ptr):m_ptr(ptr), count(new int(1))
+    {
+
     }
 
-    SharedPointer(T * ptr) : m_ptr(ptr), m_count(new unsigned int(1))
-        {
-        }
+    //Copy Constructor
+    SharedPointer(const SharedPointer& p): m_ptr(p.m_ptr),
+    count(p.count)
+    {
+        (*count)++;
+    }
+
+    //Copy assigment operator
+    SharedPointer& operator=(const SharedPointer& p )
+    {
+         if (this != &p) {
+            *count = *count> 0 ? *count-1: *count;
+            if(*count == 0)
+            {
+                delete count;
+                delete m_ptr;
+            }
+        m_ptr = p.m_ptr;
+        count = p.count;
+        (*count)++;
+         }
+         return *this;
+    }
+
+
+    //Move Constructor
+    SharedPointer(SharedPointer&& p): m_ptr(p.m_ptr),
+    count(p.count)
+    {
+         p.m_ptr = nullptr;
+         p.count = nullptr;
+    }
+
+    //Move Assigment operator
+    SharedPointer operator=(SharedPointer&& p)
+    {
+        if (this != &p) {
+            (*count)--;
+            if(*count == 0)
+            {
+                delete count;
+                delete m_ptr;
+            }
+            m_ptr = p.m_ptr;
+            count = p.count;
+            p.m_ptr = nullptr;
+            p.count = nullptr;
 
-    SharedPointer(const SharedPointer& sharedPtr){ // Copy constructor
-        this->m_ptr = sharedPtr.m_ptr;
-        this->m_count = sharedPtr.m_count;
-        if(nullptr != sharedPtr.m_ptr)
-        {
-            (*this->m_count)++;
         }
-        
+        return *this;
     }
-    
-    SharedPointer& operator=(const SharedPointer& sharedPtr) //Copy assignment operator
+
+    // Destructor 
+    ~SharedPointer()
     {
-        this->m_ptr = sharedPtr.m_ptr;
-        this->m_count = sharedPtr.m_count;
-        if(nullptr != sharedPtr.m_ptr)
+        (*count)--; 
+        if(*count == 0 )
         {
-            (*this->m_count)++;
+            delete m_ptr;
+            delete count;
         }
     }
     
-    T* operator->() const
-    {
-        return this->m_ptr;
-    }
+    //Dereferencing
+     T& operator *()
+     {
+        return *m_ptr;
+     }
+
+
+     //Pointing
+     T* operator->()
+     {
+        return m_ptr;
+     }
+ 
+     //Return use_count
+     int  use_count() const{
+        return *count;
+     }
     
+    //calculate unique
+     bool unique() const{
+        return *count == 1;
+     }
+
+     //Reset
+     void reset() {
```

Comment:
In your shared-ptr code, try this:

Sp a = .....

Sp b = std::move(a);

a.reset( );

Your program will crash. It should not crash.

Enable address sanitizer while running your code for capturing memory issues with detailed errors...

Reply:
resolved

Reply:
Not resolved properly.

In your shared-ptr code, try this:

Sp a = .....

Sp b = std::move(a);

**a.use_count( );**

Your program will crash. It should not crash.

Enable address sanitizer while running your code for capturing memory issues with detailed errors...